python/rikai/spark/sql/codegen/mlflow_registry.py

```diff
@@ -0,0 +1,303 @@
+        self._artifact = None
+        self.tracking_uri = tracking_uri
+
+    @property
+    def run_id(self):
+        return self._run.info.run_id
+
+    @property
+    def start_time(self):
+        return self._run.info.start_time
+
+    @property
+    def end_time(self):
+        return self._run.info.end_time
+
+    def load_model(self) -> Any:
+        """Load the model artifact specified in this spec"""
+        old_uri = mlflow.get_tracking_uri()
+        try:
+            mlflow.set_tracking_uri(self.tracking_uri)
+            return getattr(mlflow, self.flavor).load_model(self.uri)
+        finally:
+            mlflow.set_tracking_uri(old_uri)
+
+    def _run_to_spec_dict(self):
+        """Convert the Run into a ModelSpec
+
+        Returns
+        -------
+        spec: Dict[str, Any]
+        """
+        tags = self._run.data.tags
+        spec = {
+            "version": tags["rikai.spec.version"],
+            "schema": tags["rikai.output.schema"],
+            "model": {"flavor": tags["rikai.model.flavor"]},
+        }
+
+        # transforms
+        transforms = {}
+        if "rikai.transforms.pre" in tags:
+            transforms["pre"] = tags["rikai.transforms.pre"]
+        if "rikai.transforms.post" in tags:
+            transforms["post"] = tags["rikai.transforms.post"]
+        if len(transforms) > 0:
+            spec["transforms"] = transforms
+
+        # options
+        options = dict(self._run.data.params or {})  # for training
+        for key, value in tags.items():
+            key = key.lower().strip()
+            if key.startswith("rikai.option."):
+                sub_len = len("rikai.option.")
+                options[key[sub_len:]] = value
+        if len(options) > 0:
+            spec["options"] = options
+
+        # uri (to the model artifact)
+        if "rikai.model.artifact_path" in tags:
+            spec["model"]["uri"] = os.path.join(
+                "runs:/",
+                self._run.info.run_id,
+                tags["rikai.model.artifact_path"],
+            )
+        return spec
+
+
+def codegen_from_run(
+    spark: SparkSession,
+    run: mlflow.entities.Run,
+    tracking_uri: str,
+    name: Optional[str] = None,
+    options: Optional[Dict[str, str]] = None,
+) -> str:
+    """Generate code from an MLFlow runid
+
+    Parameters
+    ----------
+    spark : SparkSession
+        A live spark session
+    run : mlflow.entities.Run
+        the mlflow run that produced the model we want to register
+    tracking_uri: str
+        The mlflow tracking uri
+    name : str
+        The name of the model in the catalog
+    options : dict
+        Optional parameters passed to the model.
+
+    Returns
+    -------
+    str
+        Spark UDF function name for the generated data.
+    """
+    try:
+        spec = MlflowModelSpec(run, tracking_uri, options=options)
+    except Exception as e:
+        to_spec_msg = (
+            "Could not create well-formed ModelSpec from run {}."
```

Comment:
do we expect this is the only exception (spec formation?) should we differentiate exceptions?

Reply:
what did you have in mind?

Reply:
for example, is it possible to make connection in `__init__` and raises NetworkException / PermissionDenied or some sort.

Reply:
for now i'll just let individual exceptions bubble up?